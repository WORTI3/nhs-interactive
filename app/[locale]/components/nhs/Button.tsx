import React, { ReactNode } from 'react';

interface ButtonProps {
  id?: string;
  className?: string;
  type?: "submit" | "button" | "reset";
  secondary?: boolean;
  reverse?: boolean;
  preventDoubleClick?: boolean;
  as?: 'button' | 'a';
  href?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void;
  disabled?: boolean;
  children: ReactNode;
}

/**
 * Represents a customizable button component.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {string} [props.id] - The unique identifier for the button.
 * @param {string} [props.className] - Additional class names to be applied to the button.
 * @param {"submit" | "button" | "reset"} [props.type="button"] - The type of the button (submit, button, reset).
 * @param {boolean} [props.secondary=false] - Determines if the button is of secondary style.
 * @param {boolean} [props.reverse=false] - Determines if the button is of reverse style.
 * @param {boolean} [props.preventDoubleClick=false] - Determines if double-click prevention should be applied.
 * @param {'button' | 'a'} [props.as='button'] - The type of element to render (button or anchor).
 * @param {string} [props.href] - The URL to navigate to if the button is rendered as an anchor.
 * @param {(event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void} [props.onClick] - The callback function for the click event.
 * @param {boolean} [props.disabled] - Determines if the button is disabled.
 * @param {ReactNode} props.children - The content of the button.
 * @returns {ReactNode} The rendered button component.
 */
const Button: React.FC<ButtonProps> = ({
  id,
  className,
  type = "button",
  secondary = false,
  reverse = false,
  preventDoubleClick = false,
  as = 'button',
  href,
  onClick,
  disabled,
  children
}) => {
  const buttonClasses = `nhsuk-button ${secondary ? 'nhsuk-button--secondary' : ''} ${reverse ? 'nhsuk-button--reverse' : ''} ${className || ''}`;

  const handleClick = (event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => {
    if (preventDoubleClick) {
      // @ts-ignore
      event.currentTarget.disabled = true;
    }
    if (onClick) {
      onClick(event);
    }
  };

  if (as === 'a') {
    return (
      <a
        id={id}
        className={buttonClasses}
        href={href}
        onClick={handleClick}
        // @ts-ignore
        disabled={disabled}
      >
        {children}
      </a>
    );
  }

  return (
    <button
      id={id}
      className={buttonClasses}
      type={type}
      data-module="nhsuk-button"
      data-prevent-double-click={preventDoubleClick}
      onClick={handleClick}
      disabled={disabled}
    >
      {children}
    </button>
  );
};

export default Button;
