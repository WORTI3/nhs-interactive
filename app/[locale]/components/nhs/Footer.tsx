"use client";

import { Footer as NHSFooter } from "nhsuk-react-components";

export const Footer = () => {
  return (
    <NHSFooter>
      <NHSFooter.List>
        <NHSFooter.ListItem href="https://github.com/worti3" target="_blank" rel="noreferrer" aria-label="Github Profile">Github</NHSFooter.ListItem>
      </NHSFooter.List>
      {/* <NHSFooter.Copyright>&copy; </NHSFooter.Copyright> */}
    </NHSFooter>
  );
}
