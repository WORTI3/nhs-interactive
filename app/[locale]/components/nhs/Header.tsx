"use client";

import { Header as NHSHeader } from "nhsuk-react-components";
import LanguageSelector from "../LanguageSelector";
import { useTranslations } from "next-intl";

export const Header = ({locale}: {locale: string}) => {
  const translate = useTranslations('header');
  return (
    <>
      <NHSHeader transactional>
        <NHSHeader.Container>
          <NHSHeader.Logo href="/" />
          <NHSHeader.ServiceName href="/">{translate('title')}</NHSHeader.ServiceName>
          <LanguageSelector translate={translate} locale={locale} />
        </NHSHeader.Container>
      </NHSHeader>
    </>
  );
}
