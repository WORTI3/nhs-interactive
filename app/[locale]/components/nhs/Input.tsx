import { classNames } from '../../../utils';
import { parseAsString, useQueryState } from 'next-usequerystate';
import React from 'react';

interface InputProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  id?: string;
  label?: string;
  hint?: string;
  error?: string | null;
}

/**
 * Represents a customisable input component.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {string} [props.id] - The unique identifier for the input.
 * @param {string} [props.label] - The label for the input.
 * @param {string} [props.hint] - The hint for the input.
 * @param {string | null} [props.error] - The error message for the input.
 * @param {ReactNode} props.children - The content of the input.
 * @returns {ReactNode} The rendered input component.
 */
const Input: React.FC<InputProps> = ({
  id,
  label,
  hint,
  error,
  ...restProps
}) => {
  const [rows, setRows] = React.useState(1);
  const [query, setQuery] = useQueryState('q', parseAsString.withDefault(""));
  const [showText, setShowText] = React.useState(false);
  const handleChange = (event: { target: { value: React.SetStateAction<string>; }; }) => {
    if (event.target.value.length >= 50 && query.length >= 50) {
      setShowText(true);
      setRows(3);
    } else {
      setShowText(false);
      // dont set rows here
    }
    
    setQuery(event.target.value);
  };

  const inputContainerClasses = `nhsuk-character-count nhsuk-form-group ${error ? 'nhsuk-form-group--error' : ''}`;
  const textBoxClasses = `nhsuk-textarea nhsuk-js-character-count ${error ? 'nhsuk-textarea--error' : ''}`;
  
  return (
    <div className={inputContainerClasses}>
      {label && <label className="nhsuk-label" htmlFor={id}>{label}</label>}
      {hint && <span className="nhsuk-hint">{hint}</span>}
      {error && <span className="nhsuk-error-message">{error}</span>}
      <textarea 
        id={id}
        value={query}
        className={textBoxClasses} 
        rows={rows}
        maxLength={100}
        onChange={handleChange} 
        {...restProps}></textarea>
    
      { showText &&
        <div className={classNames((query.length > 95) ? "nhsuk-error-message" : "nhsuk-hint")} id="exceeding-info">
          You have {100 - query.length} characters remaining
        </div>
      }
    </div>
  );
};

export default Input;
