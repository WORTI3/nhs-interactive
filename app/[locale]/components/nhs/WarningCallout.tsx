import React, { ReactNode, ChangeEvent } from 'react';

interface WarningCalloutProps {
  className?: string;
  children: ReactNode;
}

/**
 * Represents a warning callout component.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {string} [props.className] - Additional class names to be applied to the warning callout.
 * @param {ReactNode} props.children - The content of the warning callout.
 * @returns {ReactNode} The rendered warning callout component.
 */
export const WarningCallout: React.FC<WarningCalloutProps> = ({ className, children }) => {
  const containerClasses = `${className || ''} nhsuk-warning-callout`;

  return (
    <div className={containerClasses}>
      {children}
    </div>
  );
};

interface WarningCalloutLabelProps {
  children: ReactNode;
}

/**
 * Represents a label within a warning callout.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {ReactNode} props.children - The content of the label.
 * @returns {ReactNode} The rendered label component.
 */
export const WarningCalloutLabel: React.FC<WarningCalloutLabelProps> = ({ children }) => (
  <h2 className="nhsuk-warning-callout__label">
    <span role="text">
      <span className="nhsuk-u-visually-hidden">Important: </span>
      {children}
    </span>
  </h2>
);

interface WarningCalloutTextProps {
  children: ReactNode;
}

/**
 * Can be removed in future.. relays children given in a p tag.
 * 
 * @component
 * @param {object} props - The properties of the component.
 * @returns {ReactNode} The rendered text component.
 */
export const WarningCalloutText: React.FC<WarningCalloutTextProps> = ({ children }) => {
  return <p>{children}</p>;
};
