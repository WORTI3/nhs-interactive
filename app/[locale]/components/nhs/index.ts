export { default as Button } from './Button';
export { default as Input } from './Input';
export { Header } from './Header';
export { Footer } from './Footer';
export { wrapper as Wrapper } from './wrapper';
export * from './WarningCallout';