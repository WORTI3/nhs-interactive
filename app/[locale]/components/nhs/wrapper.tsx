// not a true react component
export function wrapper(content: any) {
  return (
    <div className="nhsuk-width-container">
      <main className="nhsuk-main-wrapper" id="maincontent" role="main">
        <div className="nhsuk-grid-row">
          <div className="nhsuk-grid-column-two-thirds">
            { content }
          </div>
        </div>
      </main>
    </div>
  );
}
