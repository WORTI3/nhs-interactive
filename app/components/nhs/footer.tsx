"use client";

import { Footer as NHSFooter } from "nhsuk-react-components";

export const Footer = () => {
  return (
    <NHSFooter>
      <NHSFooter.List>
        <NHSFooter.ListItem href="/">back to top</NHSFooter.ListItem>
      </NHSFooter.List>
      <NHSFooter.Copyright>&copy; </NHSFooter.Copyright>
    </NHSFooter>
  );
}