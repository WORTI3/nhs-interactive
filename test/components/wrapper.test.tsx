import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { wrapper } from '../../app/[locale]/components/nhs/wrapper';
import { load } from 'cheerio';

describe('wrapper', () => {
  test('rendered component should match snapshot', () => {
    const { container } = render(<>{ wrapper('content', false) }</>);
    expect(container).toMatchSnapshot();
  });

  test('should render the component with the correct links & elements', () => {
    const { container } = render(<>{ wrapper('content', false) }</>);
    const html = container.innerHTML;
    const $ = load(html);
    const nhsContainer = $('.nhsuk-width-container');
  
    expect(nhsContainer.text()).toStrictEqual('content');
    expect(nhsContainer.find('p').length).toBe(0);
  });
});
