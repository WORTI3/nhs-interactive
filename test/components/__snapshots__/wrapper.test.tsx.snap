// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapper rendered component should match snapshot 1`] = `
<div>
  <div
    class="nhsuk-width-container"
  >
    <main
      class="nhsuk-main-wrapper"
      id="maincontent"
      role="main"
    >
      <div
        class="nhsuk-grid-row"
      >
        <div
          class="nhsuk-grid-column-three-quarters mt-5 mb-10"
        >
          content
        </div>
      </div>
    </main>
  </div>
</div>
`;
